### ISSUE TITLE
[GPU] Severe Gradient Precision Loss in Transpose-Pad-Conv2D Chain (Fixed in 2.17, Requesting Root Cause)

### ISSUE THREAD

**Issue by XiaoHuaOvo-tech** [tensorflow/tensorflow#107496]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

TF2.3.0 and TF2.17.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

TF2.3.0 python=3.7.12 and TF2.17.0 python=3.10.19

### CUDA/cuDNN version

TF2.3.0 CUDA=10.1 and TF2.17.0 CUDA=12.3

### GPU model and memory

Tesla V100

### Current behavior?

Hi TensorFlow Developers,

While performing Gradient Metamorphic Testing (GMR) on TensorFlow convolution gradients, we used a minimal reproduction script on GPU with `tf.nn.conv2d`, `data_format="NCHW"`, and explicit padding. We observed a deterministic violation of linearity in the convolution filter gradient (gW) in TensorFlow 2.3.0.

Test setup (minimal and deterministic):

Forward operator: `tf.nn.conv2d` on GPU with `data_format="NCHW"`, using `padding="VALID"` after manual padding
Gradient injection: explicitly injected output gradient `dY = ones_like(output)`
Linearity check: compare `gW(alpha * dY)` with `alpha * gW(dY)`, where the scaling factor is fixed at `alpha = 5.0`
Inputs, convolution kernels, and random seeds are all fixed; no randomness is involved during the test

Metric used (v_R):
We measure the maximum normalized deviation between two tensors A and B as follows:

r_i = |A_i - B_i| / (atol + rtol * max(|A_i|, |B_i|))
v_R = max_i r_i

where `atol = 1e-5` and `rtol = 1e-3`.
Intuitively, `v_R ≈ 1` indicates that the result is close to the numerical tolerance boundary, while values significantly larger than 1 usually indicate deviations beyond reasonable floating-point error.

Observed results:

TensorFlow 2.3.0:
 `v_R ≈ 26` (on the order of 10–30, stably reproducible across multiple runs), which is well beyond numerical tolerance and indicates a severe linearity violation.
TensorFlow 2.17.0** (same script, same configuration):
 `v_R ≈ 0.010` (absolute difference ≈ 4e-4), which is consistent with the expected FP32 numerical error on GPU.

Based on the above comparison, we believe that the linearity issue in convolution filter backpropagation on the GPU + NCHW execution path has been fixed in newer TensorFlow versions.

We would like to ask:

1. Is this a known issue?
2. From which TensorFlow version was the fix introduced, or is there a related issue / PR that we could reference?

This information would help us mark this behavior as a known issue in our testing framework.

We have attached the minimal reproducible script and would greatly appreciate a response at your convenience. Thank you very much for your time and help.

Best regards,
Dong Hao

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import numpy as np
import os


os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'
gpus = tf.config.list_physical_devices('GPU')
if gpus:
    try:
        for gpu in gpus:
            tf.config.experimental.set_memory_growth(gpu, True)
        print(f"Running on GPU: {gpus[0].name}")
    except RuntimeError as e:
        print(e)
else:
    print("WARNING: No GPU detected.")


def compute_vR(A, B, atol=1e-5, rtol=1e-3):
    A_flat = np.asarray(A).flatten()
    B_flat = np.asarray(B).flatten()
    
    diff = np.abs(A_flat - B_flat)
    scale = np.maximum(np.abs(A_flat), np.abs(B_flat))
    tolerance = atol + rtol * scale
    
    eps = np.finfo(np.float64).tiny
    r_i = diff / (tolerance + eps)
    
    v_R = np.max(r_i)
    
    worst_idx = np.argmax(r_i)
    worst_info = {
        'vR': v_R,
        'idx': worst_idx,
        'diff': diff[worst_idx],
        'val_A': A_flat[worst_idx],
        'val_B': B_flat[worst_idx]
    }
    return v_R, worst_info


def reproduce_bug():
    print("\n=== Starting Reproduction (Strict Gradient Injection Mode) ===")

    # --- Seed 85  ---
    # NCHW, Explicit Padding, Standard Conv (ignoring groups=3 tag as per code analysis)
    INPUT_SHAPE = (10, 3, 32, 32)
    FILTERS = 6
    KERNEL_SIZE = (5, 5)
    STRIDES_VAL = 1  # s=(1,1)
    PAD_VAL = 4      # p=4 (Explicit)
    
    SEED = 85
    tf.random.set_seed(SEED)
    np.random.seed(SEED)

    x_val = tf.random.uniform(shape=INPUT_SHAPE, minval=-1, maxval=1, dtype=tf.float32)
    x_var = tf.Variable(x_val)

    k_val_nchw = tf.random.normal((FILTERS, INPUT_SHAPE[1], KERNEL_SIZE[0], KERNEL_SIZE[1]), dtype=tf.float32)
    k_var = tf.Variable(k_val_nchw)
    
    strides = [1, STRIDES_VAL, STRIDES_VAL, 1] # NHWC format strides
    dilations = [1, 1, 1, 1]

    def run_conv_and_grad(input_data, kernel_data, upstream_grad=None):
        with tf.GradientTape(persistent=True) as tape:
            tape.watch(input_data)
            tape.watch(kernel_data)

            # 原生 NCHW 格式 - 直接在 NCHW 格式上进行 Padding
            # Input: [N, C, H, W]
            padded_data_nchw = tf.pad(input_data,
                                      [[0, 0], [0, 0], [PAD_VAL, PAD_VAL], [PAD_VAL, PAD_VAL]],
                                      mode='CONSTANT')

            # Kernel 已经是 NCHW 格式: [Out, In, kH, kW]
            # 对于 NCHW 的 conv2d，kernel 应该是 [kH, kW, In, Out]
            kernel_hwio = tf.transpose(kernel_data, [2, 3, 1, 0])

            # 原生 NCHW 卷积 - 使用 data_format='NCHW'
            output = tf.nn.conv2d(
                input=padded_data_nchw,
                filters=kernel_hwio,
                strides=[1, 1, STRIDES_VAL, STRIDES_VAL],  # NCHW 格式的 strides: [1, 1, stride_h, stride_w]
                padding='VALID',  # 已经手动 Pad 了
                data_format='NCHW',  # 关键：指定 NCHW 格式
                dilations=[1, 1, 1, 1]
            )

        # --- 显式梯度注入 (Explicit Gradient Injection) ---
        # 对应 utils.py 中 generate_gradient_G(mode="G0") -> tf.ones
        if upstream_grad is None:
            G = tf.ones_like(output)
        else:
            G = upstream_grad
            
        # 计算梯度 (传入 output_gradients)
        # 这就是你要的“全1梯度注入”
        gW = tape.gradient(output, kernel_data, output_gradients=G)
        
        del tape
        return gW, output

    print("Running GMR Check...")

    # Run 1: 原始执行
    # 获取 output 是为了生成 G (虽然这里是全1，但逻辑上需要 shape)
    _, out_1 = run_conv_and_grad(x_var, k_var, upstream_grad=None)
    
    # 生成全1梯度 G
    G0 = tf.ones_like(out_1)
    
    # GMR 线性变换: G_scaled = alpha * G
    alpha = 5.0
    G_scaled = alpha * G0
    
    # 计算原始梯度 gW (注入 G0)
    gW_orig, _ = run_conv_and_grad(x_var, k_var, upstream_grad=G0)
    
    # 计算变换后梯度 gW_scaled (注入 G_scaled)
    # 对应 GMR_1 函数中的: gX_scaled, gW_scaled, ... = compute_conv_gradients(..., G_scaled)
    gW_scaled, _ = run_conv_and_grad(x_var, k_var, upstream_grad=G_scaled)
    
    # --- 验证 (Validation) ---
    if gW_orig is None or gW_scaled is None:
        print("Error: Gradients are None!")
        return

    # 理论上: gW_scaled 应等于 alpha * gW_orig
    # 你的 GMR_1 实现: compute_vR(gW_scaled, alpha * gW)
    v_R, info = compute_vR(gW_scaled.numpy(), (alpha * gW_orig).numpy())
    
    print(f"\n[Result] v_R: {v_R:.4f}")
    print(f"Details: {info}")
    
    if v_R > 1.0:
        print("\n❌ VIOLATION FOUND!")
        if v_R > 100.0:
            print("🚨 CATASTROPHIC FAILURE (Matches Seed 85)")
    else:
        print("\n✅ Check Passed")

if __name__ == "__main__":
    reproduce_bug()
```

### Relevant log output

```shell
`v_R ≈ 26` (on the order of 10–30, stably reproducible across multiple runs), which is well beyond numerical tolerance and indicates a severe linearity violation.
TensorFlow 2.17.0** (same script, same configuration):
 `v_R ≈ 0.010` (absolute difference ≈ 4e-4), which is consistent with the expected FP32 numerical error on GPU.
```

**Comment by XiaoHuaOvo-tech** [comment]
Subject: Detailed Reproduction Data & Regression Confirmation
Hi there, thanks for looking into this.
To clarify the context of this report: We are conducting a longitudinal study on TensorFlow's numerical stability.
1. Confirmation of the Bug in Older Versions : We have confirmed that this Transpose-Pad-Conv2D chain causes  gradient errors in TF 2.3.0 through TF 2.10.0.
TF 2.3.0 (NCHW ): v_R ≈ 26 
TF 2.10.0 (NCHW ): v_R ≈ 33
2. Confirmation of Fix in TF 2.17.0: We verified that in TF 2.17.0, the same reproduction script yields v_R < 1.0 (Pass),v_R ≈ 0.0113. This indicates the underlying issue (likely in the Auto-diff graph optimization for Transpose/Pad ops) has been resolved in recent versions.
3. Request: Since this was a severe "silent bug" in previous versions:
Could you help identify which specific PR or commit fixed this issue?
Thank you!

**Comment by nileshau-afk** [comment]
Hi @XiaoHuaOvo-tech, 

Thank you for your patience, and apologies for the late reply. The gradient precision loss you observed in TensorFlow 2.3.0 for `tf.nn.conv2d` with `data_format="NCHW"` and explicit padding on GPU was indeed a bug related to the backend implementation of convolution gradients. This issue was resolved in later TensorFlow versions, as you've confirmed with 2.17.0. Pinpointing the exact commit or PR for such a specific, older bug can be challenging due to the continuous evolution of the XLA and GPU backends, which often involve numerous interdependent changes rather than a single atomic fix. It's likely that improvements to the `NCHW` data format handling and gradient computations within the XLA compiler and cuDNN integration contributed to this resolution between versions 2.3.0 and 2.17.0. 

We recommend using a recent stable TensorFlow version (e.g., 2.17.0 or newer) to benefit from these and other performance/correctness improvements. Please let us know if upgrading to a newer version addresses your needs for this specific issue.

**Comment by djhdusjbd-eng** [comment]
I’m interested—do you have other similar opportunities?
